File: backend/app/api/usage.py
from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from backend.app.api.dependencies import get_repositories
from backend.app.services import (
    UsageRequestNotFoundError,
    UsageRequestQuery,
    UsageService,
    UsageValidationError,
)
from backend.app.storage import StorageRepositories

router = APIRouter(prefix="/api/usage", tags=["usage"])
RepositoriesDep = Depends(get_repositories)


class UsageSummaryResponse(BaseModel):
    request_count: int
    total_tokens: int
    input_tokens: int
    cache_read_tokens: int
    output_tokens: int
    success_count: int
    failed_count: int
    avg_duration_ms: int


class UsageTrendPoint(BaseModel):
    time: str
    request_count: int
    input_tokens: int
    cache_read_tokens: int
    output_tokens: int
    total_tokens: int
    failed_count: int


class UsageTrendResponse(BaseModel):
    points: list[UsageTrendPoint]


class UsageRequestListResponse(BaseModel):
    list: list[dict[str, Any]]
    total: int
    page: int
    page_size: int


class UsageRequestDetailResponse(BaseModel):
    request: dict[str, Any]
    trace: dict[str, Any] | None
    events: list[dict[str, Any]]


@router.get("/summary", response_model=UsageSummaryResponse)
def get_usage_summary(
    start_time: str | None = None,
    end_time: str | None = None,
    model: str | None = None,
    repositories: StorageRepositories = RepositoriesDep,
) -> UsageSummaryResponse:
    return UsageSummaryResponse(
        **_service(repositories).get_summary(
            start_time=start_time,
            end_time=end_time,
            model=model,
        )
    )


@router.get("/trend", response_model=UsageTrendResponse)
def get_usage_trend(
    start_time: str | None = None,
    end_time: str | None = None,
    bucket: Literal["hour", "day"] = "day",
    timezone_offset_minutes: int = 0,
    model: str | None = None,
    repositories: StorageRepositories = RepositoriesDep,
) -> UsageTrendResponse:
    try:
        points = _service(repositories).get_trend(
            start_time=start_time,
            end_time=end_time,
            model=model,
            bucket=bucket,
            timezone_offset_minutes=timezone_offset_minutes,
        )
    except UsageValidationError as exc:
        raise _bad_request("invalid_usage_query", str(exc)) from exc
    return UsageTrendResponse(points=points)


@router.get("/requests", response_model=UsageRequestListResponse)
def list_usage_requests(
    start_time: str | None = None,
    end_time: str | None = None,
    model: str | None = None,
    status_filter: str | None = Query(default=None, alias="status"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    repositories: StorageRepositories = RepositoriesDep,
) -> UsageRequestListResponse:
    try:
        result = _service(repositories).list_requests(
            UsageRequestQuery(
                start_time=start_time,
                end_time=end_time,
                model=model,
                status=status_filter,
                page=page,
                page_size=page_size,
            )
        )
    except UsageValidationError as exc:
        raise _bad_request("invalid_usage_query", str(exc)) from exc
    return UsageRequestListResponse(**result)


@router.get("/requests/{request_id}", response_model=UsageRequestDetailResponse)
def get_usage_request_detail(
    request_id: str,
    repositories: StorageRepositories = RepositoriesDep,
) -> UsageRequestDetailResponse:
    try:
        detail = _service(repositories).get_request_detail(request_id)
    except UsageRequestNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "usage_request_not_found", "message": str(exc), "details": {}},
        ) from exc
    return UsageRequestDetailResponse(**detail)


def _service(repositories: StorageRepositories) -> UsageService:
    return UsageService(repositories)


def _bad_request(code: str, message: str) -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail={"code": code, "message": message, "details": {}},
    )

File: desktop/src/runtime/usage.ts
import type {
  UsageBucket,
  UsageRequestDetail,
  UsageRequestListResponse,
  UsageRequestStatus,
  UsageSummary,
  UsageTrendResponse,
} from "@/types/protocol";

import type { HttpClient } from "./httpClient";

export interface UsageQueryOptions {
  startTime?: string;
  endTime?: string;
  model?: string;
}

export interface UsageTrendOptions extends UsageQueryOptions {
  bucket?: UsageBucket;
  timezoneOffsetMinutes?: number;
}

export interface UsageRequestListOptions extends UsageQueryOptions {
  status?: UsageRequestStatus;
  page?: number;
  pageSize?: number;
}

export interface UsageRuntime {
  getSummary(options?: UsageQueryOptions): Promise<UsageSummary>;
  getTrend(options?: UsageTrendOptions): Promise<UsageTrendResponse>;
  listRequests(options?: UsageRequestListOptions): Promise<UsageRequestListResponse>;
  getRequestDetail(requestId: string): Promise<UsageRequestDetail>;
}

export function createUsageRuntime(http: HttpClient): UsageRuntime {
  return {
    getSummary(options = {}) {
      return http.request<UsageSummary>(`/api/usage/summary${usageQuery(options)}`);
    },
    getTrend(options = {}) {
      return http.request<UsageTrendResponse>(
        `/api/usage/trend${usageQuery({ ...options, bucket: options.bucket })}`,
      );
    },
    listRequests(options = {}) {
      return http.request<UsageRequestListResponse>(`/api/usage/requests${usageQuery(options)}`);
    },
    getRequestDetail(requestId) {
      return http.request<UsageRequestDetail>(`/api/usage/requests/${encodeURIComponent(requestId)}`);
    },
  };
}

function usageQuery(options: UsageQueryOptions & Partial<UsageTrendOptions & UsageRequestListOptions>) {
  const params = new URLSearchParams();
  appendParam(params, "start_time", options.startTime);
  appendParam(params, "end_time", options.endTime);
  appendParam(params, "model", options.model);
  appendParam(params, "bucket", options.bucket);
  appendParam(params, "timezone_offset_minutes", options.timezoneOffsetMinutes);
  appendParam(params, "status", options.status);
  appendParam(params, "page", options.page);
  appendParam(params, "page_size", options.pageSize);
  const query = params.toString();
  return query ? `?${query}` : "";
}

function appendParam(params: URLSearchParams, key: string, value: string | number | undefined | null) {
  if (value === undefined || value === null || value === "") {
    return;
  }
  params.set(key, String(value));
}
